storage.go

```diff
@@ -325,6 +326,84 @@ func (s *Storage) keyList() (map[string]bool, uint64, error) {
 	}
 	return keyList, rev, nil
 }
+
+// keyListAdd adds the given key to the keylist
+func (s *Storage) keyListAdd(key string) error {
+	keyList, keyListRev, err := s.keyList()
+	if err != nil && !errNotFound(err) {
+		return err
+	}
+	if keyList == nil {
+		keyList = make(map[string]bool)
+	}
+	// If the key is already in the keylist there's nothing to do.
+	if _, exists := keyList[key]; exists {
+		return nil
+	}
+	keyList[key] = true
+	bytes, err := json.Marshal(keyList)
+	if err != nil {
+		return errors.AddContext(err, "failed to serialise the new key list")
+	}
+	err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
+	if err != nil {
+		return errors.AddContext(err, "failed to store the key list")
+	}
+	return nil
+}
+
+// keyListDelete deletes the given key from the keylist
+func (s *Storage) keyListDelete(key string) error {
+	keyList, keyListRev, err := s.keyList()
+	if err != nil && !errNotFound(err) {
+		return err
+	}
+	// If the keylist is empty there's nothing to do.
+	if keyList == nil {
+		return nil
+	}
+	// If the key is not in the keylist there's nothing to do.
+	if _, exists := keyList[key]; !exists {
+		return nil
+	}
+	delete(keyList, key)
+	bytes, err := json.Marshal(keyList)
+	if err != nil {
+		return errors.AddContext(err, "failed to serialise the new key list")
+	}
+	err = s.SkyDB.Write(bytes, s.KeyListDataKey, keyListRev+1)
+	if err != nil {
+		return errors.AddContext(err, "failed to store the key list")
+	}
+	return nil
+}
+
+// writeItem is a helper that writes a new item to SkyDB.
+func (s *Storage) writeItem(pk string, contents []byte, rev uint64) error {
+	item := Item{
+		PrimaryKey:  pk,
+		Contents:    contents,
+		LastUpdated: time.Now().UTC(),
+	}
+	bytes, err := json.Marshal(item)
+	if err != nil {
+		return errors.AddContext(err, "failed to marshal the item record")
+	}
+	dataKey := crypto.HashBytes([]byte(item.PrimaryKey))
```

Comment:
What is the key, a public key? I find it strange that we calculate the data key by hashing a key.

Reply:
The item's primary key is basically its identifier, its name. Think SQL primary keys, not cryptography keys.  
The reason we form the dataKey by hashing the primary key is this is a direct one-to-one mapping between the item's name (primary key) and where its value is stored.